### ISSUE TITLE
Where should the grad2 symbol in _perlin.so come from?

### ISSUE THREAD

**Issue by klausman** [caseman/noise#19]
```
$ pip2 install --user --force-reinstall --ignore-installed --no-binary :all:  noise 
Collecting noise
  Using cached noise-1.2.2.tar.gz
Skipping bdist_wheel for noise, due to binaries being disabled for it.
Installing collected packages: noise
  Running setup.py install for noise ... done
Successfully installed noise-1.2.2
$ python2
Python 2.7.13 (default, May  8 2017, 19:39:57) 
[GCC 6.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import noise
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/klausman/.local/lib64/python2.7/site-packages/noise/__init__.py", line 12, in <module>
    from . import _perlin, _simplex
ImportError: /home/klausman/.local/lib64/python2.7/site-packages/noise/_perlin.so: undefined symbol: grad2
>>> 
```

**Comment by klausman** [comment]
This is caused by various functions being declared inline (or, rather: removing the inline declaration makes it compile, even though `fast_cos` and `fast_sin` are then still missing. I dropped those to just `cos` and `sin`, respectively).

**Comment by happysmash27** [comment]
Holy cow, that actually worked! Just remove all of the words "newline" in the function declarations in the _perlin.c and _simplex.c files, and in _simplex.c, replace all instances of fast_cos and fast_sin with cos and sin, respectively, and, somehow, it works... 

Maybe this should have a pull request?

**Comment by klausman** [comment]
The last commit on this repo was in January 2016. I doubt a PR would speed things up.

**Comment by caseman** [comment]
What platform is this? There's some inline handling for windows/VC in _noise.h, but I've never had any issue with gcc handling inline.

**Comment by klausman** [comment]
This was a recent gcc on x86_64 linux. The problem wasn't the inlining itself, but that the inline definition used a symbol (`grad2`) that was undefined. This might be due to not linking the right shared objects, or that said symbol was deprecated at some point.

**Comment by caseman** [comment]
Except that `grad2()` is actually defined in _perlin.c and that's the only place its referenced. Since it's declared inline there's no linking involved. 

I tried a source build with python 2.7 and 3.6 on ubuntu 16.04 and gcc 5.4 and it works as expected. I also tried it via pip using:

`pip2 install --user --force-reinstall --ignore-installed --no-binary :all:  noise`

and I was able to import it successfully. Maybe try a source built to take pip out of the equation. Create a throwaway directory and run:

```
git clone "https://github.com/caseman/noise.git" && \
        cd noise && python2.7 setup.py build && \
        cd build/lib.* && python2.7 -c "import noise; help(noise)" && \
        python2.7 test.py && gcc --version 
```

**Comment by klausman** [comment]
```
src $ git clone "https://github.com/caseman/noise.git" && \
>         cd noise && python2.7 setup.py build && \
>         cd build/lib.* && python2.7 -c "import noise; help(noise)" && \
>         python2.7 test.py && gcc --version
Cloning into 'noise'...
remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (8/8), done.
remote: Total 197 (delta 3), reused 3 (delta 1), pack-reused 188
Receiving objects: 100% (197/197), 152.97 KiB | 702.00 KiB/s, done.
Resolving deltas: 100% (110/110), done.
running build
running build_py
creating build
creating build/lib.linux-x86_64-2.7
creating build/lib.linux-x86_64-2.7/noise
copying __init__.py -> build/lib.linux-x86_64-2.7/noise
copying perlin.py -> build/lib.linux-x86_64-2.7/noise
copying shader.py -> build/lib.linux-x86_64-2.7/noise
copying shader_noise.py -> build/lib.linux-x86_64-2.7/noise
copying test.py -> build/lib.linux-x86_64-2.7/noise
running build_ext
building 'noise._simplex' extension
creating build/temp.linux-x86_64-2.7
x86_64-pc-linux-gnu-gcc -pthread -fPIC -I/usr/include/python2.7 -c _simplex.c -o build/temp.linux-x86_64-2.7/_simplex.o -funroll-loops
x86_64-pc-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,--as-needed -L. build/temp.linux-x86_64-2.7/_simplex.o -L/usr/lib64 -lpython2.7 -o build/lib.linux-x86_64-2.7/noise/_simplex.so
building 'noise._perlin' extension
x86_64-pc-linux-gnu-gcc -pthread -fPIC -I/usr/include/python2.7 -c _perlin.c -o build/temp.linux-x86_64-2.7/_perlin.o -funroll-loops
x86_64-pc-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,--as-needed -L. build/temp.linux-x86_64-2.7/_perlin.o -L/usr/lib64 -lpython2.7 -o build/lib.linux-x86_64-2.7/noise/_perlin.so
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "noise/__init__.py", line 12, in <module>
    from . import _perlin, _simplex
ImportError: noise/_perlin.so: undefined symbol: grad2
lib.linux-x86_64-2.7 (master=) $ gcc --version
gcc (Gentoo 8.2.0-r4 p1.5) 8.2.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

lib.linux-x86_64-2.7 (master=) $ 
```

**Comment by caseman** [comment]
Ok, interesting. I tried a build with Ubuntu 18.0.4 and gcc 8.2, with python compiled from source and it still works for me, but the gcc flags are pretty different, in particular the `-fPIC` option that generates relocatable code, which I think could be exposing this.

I think declaring all the inline functions static may make your gcc happy. I made a branch so that you can test this since I can't get mine to fail in the same way. I suspect it is something about the way gentoo does this that is different from Ubuntu. In any case give this a try:

```
git clone "https://github.com/caseman/noise.git" && \
        cd noise && git checkout static-inline && python2.7 setup.py build && \
        cd build/lib.* && python2.7 -c "import noise; help(noise)" && \
        python2.7 test.py
```

Thanks!

**Comment by klausman** [comment]
That indeed works, with one small correction: the last command should be:

`PYTHONPATH=. python2.7 ../../test.py`

Then it also passes the test suite.

**Comment by caseman** [comment]
Ok, sweet. Thanks so much for your help testing! I will merge this change in and get it published upstream.

**Comment by caseman** [comment]
The change is now in master. I am going to delete that branch since it also accidentally contains some other changes that were unrelated. Feel free to use master until a new release is published.

**Comment by funnyflowerpot** [comment]
Dear @caseman, thank you very much for fixing this problem!

The latest package version available via `pip`/PyPI is 1.2.2. Could you please upload the current version there?

I wanted to install `noise` today for Python 3 on Gentoo via `pip` and ran into this issue. I could fix it by cloning the repository and compiling manually, so I can confirm that version 1.2.3 solves the problem for me, too.

**Comment by henryflorence** [comment]
Hey, I'm experiencing this error trying to compile noise on Mint Linux under Python 2.7.17. It will compile under Python 3, however I am using other libraries which are not Python 3 compatible. I've tried deleting all inline and fast_ from _simplex.c and _perlin.c, and also compiling from caseman/noise, but the error persists.
